core/src/main/java/com/datastax/oss/driver/internal/core/cql/PagingIterableSpliterator.java

```diff
@@ -0,0 +1,167 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.datastax.oss.driver.internal.core.cql;
+
+import com.datastax.oss.driver.api.core.PagingIterable;
+import com.datastax.oss.driver.shaded.guava.common.base.Preconditions;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import java.util.Objects;
+import java.util.Spliterator;
+import java.util.Spliterators;
+import java.util.function.Consumer;
+import net.jcip.annotations.NotThreadSafe;
+
+/**
+ * A Spliterator for {@link PagingIterable} instances that splits the stream in chunks of equal
+ * size.
+ *
+ * @param <ElementT> The element type of the underlying stream.
+ */
+@NotThreadSafe
+public class PagingIterableSpliterator<ElementT> implements Spliterator<ElementT> {
+
+  /**
+   * Creates a new {@link Builder} for {@link PagingIterableSpliterator} instances.
+   *
+   * @param iterable The {@link PagingIterable} to create a spliterator for.
+   * @param <ElementT> The iterable's element type
+   * @return a new {@link Builder}.
+   */
+  @NonNull
+  public static <ElementT> Builder<ElementT> builder(@NonNull PagingIterable<ElementT> iterable) {
+    return new Builder<>(iterable);
+  }
+
+  /** The default chunk size for {@link PagingIterableSpliterator}. */
+  public static final int DEFAULT_CHUNK_SIZE = 128;
+
+  private final PagingIterable<ElementT> iterable;
+  private long estimatedSize;
+  private final int chunkSize;
+  private final int characteristics;
+
+  /**
+   * Creates a new {@link PagingIterableSpliterator} for the given iterable, with unknown size and
+   * default chunk size ({@value #DEFAULT_CHUNK_SIZE}).
+   *
+   * @param iterable The {@link PagingIterable} to create a spliterator for.
+   */
+  public PagingIterableSpliterator(@NonNull PagingIterable<ElementT> iterable) {
+    this(iterable, Long.MAX_VALUE, DEFAULT_CHUNK_SIZE);
+  }
+
+  private PagingIterableSpliterator(
+      @NonNull PagingIterable<ElementT> iterable, long estimatedSize, int chunkSize) {
+    this.iterable = Objects.requireNonNull(iterable, "iterable cannot be null");
+    this.estimatedSize = estimatedSize;
+    this.chunkSize = chunkSize;
+    if (estimatedSize < Long.MAX_VALUE) {
+      characteristics =
+          Spliterator.ORDERED
+              | Spliterator.IMMUTABLE
+              | Spliterator.NONNULL
+              | Spliterator.SIZED
+              | Spliterator.SUBSIZED;
+    } else {
+      characteristics = Spliterator.ORDERED | Spliterator.IMMUTABLE | Spliterator.NONNULL;
+    }
+  }
+
+  @Override
+  public boolean tryAdvance(Consumer<? super ElementT> action) {
+    Objects.requireNonNull(action, "action cannot be null");
+    ElementT row = iterable.one();
+    if (row == null) {
+      return false;
+    }
+    action.accept(row);
+    return true;
+  }
+
+  @Override
+  @Nullable
+  public Spliterator<ElementT> trySplit() {
```

Comment:
I wonder how do we relate to docs of `trySplit`:
```
  * An ideal {@code trySplit} method efficiently (without
     * traversal) divides its elements exactly in half, allowing
     * balanced parallel computation. Many departures from this ideal
     * remain highly effective;
```
I see that we are supporting a couple of ways to split (not only split in half)
Is it needed to give that support to a client?

Reply:
It's impossible to split a non-`SIZED` stream in two halfs of same size. The next best thing to do is to split in chunks of equal size.